File: laser-safety-lms/src/lib/progress.ts
'use client';

import { UserProgress, CourseProgress } from '@/types';

const PROGRESS_KEY = 'laser-lms-progress';
const USER_NAME_KEY = 'laser-lms-username';

function createDefaultUserProgress(): UserProgress {
  return {
    userId: '',
    courses: {},
    modules: {},
    totalTimeSpent: 0,
    streakDays: 0,
    lastActivityAt: new Date().toISOString(),
  };
}

function createDefaultCourseProgress(courseId: string): CourseProgress {
  return {
    courseId,
    status: 'available',
    progress: 0,
    completedModules: [],
    quizAttempts: 0,
  };
}

export function getProgress(): UserProgress {
  if (typeof window === 'undefined') return createDefaultUserProgress();
  try {
    const raw = localStorage.getItem(PROGRESS_KEY);
    if (!raw) return createDefaultUserProgress();
    const parsed = JSON.parse(raw);
    // Ensure all required fields exist (handle corrupted/partial data)
    return {
      ...createDefaultUserProgress(),
      ...parsed,
      courses: parsed.courses || {},
      modules: parsed.modules || {},
    };
  } catch {
    return createDefaultUserProgress();
  }
}

function saveProgress(progress: UserProgress): void {
  if (typeof window === 'undefined') return;
  localStorage.setItem(PROGRESS_KEY, JSON.stringify(progress));
}

export function getCourseProgress(courseId: string): CourseProgress {
  const progress = getProgress();
  return progress.courses[courseId] ?? createDefaultCourseProgress(courseId);
}

export function isModuleComplete(courseId: string, moduleId: string): boolean {
  return getCourseProgress(courseId).completedModules.includes(moduleId);
}

export function markModuleComplete(courseId: string, moduleId: string): void {
  const progress = getProgress();
  const course = progress.courses[courseId] ?? createDefaultCourseProgress(courseId);
  if (!course.completedModules.includes(moduleId)) {
    course.completedModules = [...course.completedModules, moduleId];
  }
  progress.courses[courseId] = course;
  progress.lastActivityAt = new Date().toISOString();
  saveProgress(progress);
}

export function saveQuizResult(courseId: string, score: number, passed: boolean): void {
  const progress = getProgress();
  const course = progress.courses[courseId] ?? createDefaultCourseProgress(courseId);
  course.quizScore = score;
  course.quizPassed = passed;
  if (passed) {
    course.completedAt = new Date().toISOString();
  }
  progress.courses[courseId] = course;
  progress.lastActivityAt = new Date().toISOString();
  saveProgress(progress);
}

export function getOverallStats(totalModules: number, totalCourses: number) {
  const progress = getProgress();
  let completedModules = 0;
  let coursesFinished = 0;

  for (const courseProgress of Object.values(progress.courses || {})) {
    completedModules += courseProgress.completedModules.length;
    if (courseProgress.quizPassed) coursesFinished++;
  }

  return {
    completedModules,
    coursesFinished,
    totalModules,
    totalCourses,
    overallPercent: totalModules > 0 ? Math.round((completedModules / totalModules) * 100) : 0,
  };
}

export function getUserName(): string {
  if (typeof window === 'undefined') return '';
  return localStorage.getItem(USER_NAME_KEY) ?? '';
}

export function setUserName(name: string): void {
  if (typeof window === 'undefined') return;
  localStorage.setItem(USER_NAME_KEY, name);
}

export function resetAllProgress(): void {
  if (typeof window === 'undefined') return;
  localStorage.removeItem(PROGRESS_KEY);
}

File: legacy/lms_data/progress.py
"""Session-state based progress tracking for Laser Safety LMS."""
import streamlit as st
from lms_data.courses import COURSES

TOTAL_MODULES = sum(len(c["modules"]) for c in COURSES)
TOTAL_COURSES = len(COURSES)


def init_progress():
    if "progress" not in st.session_state:
        st.session_state.progress = {}
    if "user_name" not in st.session_state:
        st.session_state.user_name = ""


def get_course_progress(course_id: str) -> dict:
    return st.session_state.progress.get(course_id, {"completed_modules": []})


def is_module_complete(course_id: str, module_id: str) -> bool:
    return module_id in get_course_progress(course_id).get("completed_modules", [])


def mark_module_complete(course_id: str, module_id: str):
    if course_id not in st.session_state.progress:
        st.session_state.progress[course_id] = {"completed_modules": []}
    mods = st.session_state.progress[course_id]["completed_modules"]
    if module_id not in mods:
        mods.append(module_id)


def save_quiz_result(course_id: str, score: int, total: int, passed: bool):
    if course_id not in st.session_state.progress:
        st.session_state.progress[course_id] = {"completed_modules": []}
    from datetime import datetime
    st.session_state.progress[course_id]["quiz_score"] = score
    st.session_state.progress[course_id]["quiz_total"] = total
    st.session_state.progress[course_id]["quiz_passed"] = passed
    if passed:
        st.session_state.progress[course_id]["completed_at"] = datetime.now().isoformat()


def get_overall_stats() -> dict:
    completed_modules = 0
    courses_finished = 0
    for course_prog in st.session_state.get("progress", {}).values():
        completed_modules += len(course_prog.get("completed_modules", []))
        if course_prog.get("quiz_passed"):
            courses_finished += 1
    overall_pct = round(completed_modules / TOTAL_MODULES * 100) if TOTAL_MODULES else 0
    return {
        "completed_modules": completed_modules,
        "total_modules": TOTAL_MODULES,
        "courses_finished": courses_finished,
        "total_courses": TOTAL_COURSES,
        "overall_pct": overall_pct,
    }
